### PR TITLE
Consolidated Application Config

### PR DESCRIPTION
**Pull request by nickschuch**
**Problem**

Applications have to do too much work to load configuration eg. MySQL connection details.

Applications have to load the files from a directory structure (depending on configuration) and consolidate it.

```
/etc/skpr/config/default
/etc/skpr/config/override
/etc/skpr/secret/default
/etc/skpr/secret/override
```

This is too much work for applications and exposes extra config/secrets which don't need to be.

**Solution**

Expose consolidated as a single JSON file which can be loaded once by the application.

eg.

`/etc/skpr/data/config.json`

NOTE: We cannot use the single file mount as it doesn't support live reloading.

### CODE REVIEW

File: internal/controller/app/drupal/controller.go [hunk 1]
```diff
@@ -718,14 +740,17 @@ func (r *ReconcileDrupal) Sync(log log.Logger, drupal *appv1beta1.Drupal) (appv1
...
-						generate.VolumeConfigMap(VolumeConfigOverride, configMapOverride.ObjectMeta.Name),
-						generate.VolumeSecret(VolumeSecretDefault, secretDefault.ObjectMeta.Name),
-						generate.VolumeSecret(VolumeSecretOverride, secretOverride.ObjectMeta.Name),
+						generate.VolumeSecret(VolumeConfigData, configMapData.ObjectMeta.Name),
 						generate.VolumeSecret(VolumeSecretCertificate, secretCertificate.ObjectMeta.Name),
 						generate.VolumeClaim(VolumePublic, pvcPublic.ObjectMeta.Name),
 						generate.VolumeClaim(VolumeTemporary, pvcTemporary.ObjectMeta.Name),
 						generate.VolumeClaim(VolumePrivate, pvcPrivate.ObjectMeta.Name),
+
+						// Deprecated: These will be removed in a future release.
```
Comment (nicksantamaria) [inline internal/controller/app/drupal/controller.go]: I reckon we add an `@expire` tag so we don't lose track of it. As per https://www.previousnext.com.au/blog/revisiting-technical-debt-simple-expiry-annotations

File: internal/controller/app/drupal/config.go [hunk 2]
```diff
@@ -0,0 +1,26 @@
+package drupal
+
+import (
+	"encoding/json"
+
+	corev1 "k8s.io/api/core/v1"
+)
+
+// BuildConslidatedConfig from Secrets and ConfigMaps which can be consumed by an application.
+func BuildConslidatedConfig(defaultConfig, overrideConfig *corev1.ConfigMap, defaultSecret, overrideSecret *corev1.Secret) ([]byte, error) {
```
[the remark this answers is not in the document]
Comment (nickschuch) [reply internal/controller/app/drupal/config.go]: I could make this a little more flexible / reusable for other projects eg. our workflow.

File: internal/controller/app/drupal/controller.go (hunk 1 above)
Comment (nickschuch) [reply internal/controller/app/drupal/controller.go]: Good call! Done

File: internal/controller/app/drupal/controller.go [hunk 3]
```diff
@@ -477,7 +496,7 @@ func (r *ReconcileDrupal) Sync(log log.Logger, drupal *appv1beta1.Drupal) (appv1
 		Data: make(map[string][]byte),
 	}
 
-	result, err = controllerutil.CreateOrUpdate(context.TODO(), r.Client, configMapOverride, k8ssync.ConfigMap(drupal, configMapOverride.Data, false, r.scheme))
+	result, err = controllerutil.CreateOrUpdate(context.TODO(), r.Client, configMapOverride, k8ssync.ConfigMap(drupal, configMapOverride.Data, configMapOverride.BinaryData,false, r.scheme))
```
Comment (acbramley) [inline internal/controller/app/drupal/controller.go]: Missing a space after the comma on the new param? Does that not fail linting?

**Comment by acbramley** [comment]
This is 🔥 let's get it in!

File: internal/controller/app/drupal/controller.go [hunk 4]
```diff
@@ -718,14 +741,18 @@ func (r *ReconcileDrupal) Sync(log log.Logger, drupal *appv1beta1.Drupal) (appv1
 						},
 					},
 					Volumes: []corev1.Volume{
-						generate.VolumeConfigMap(VolumeConfigDefault, configMapDefault.ObjectMeta.Name),
-						generate.VolumeConfigMap(VolumeConfigOverride, configMapOverride.ObjectMeta.Name),
-						generate.VolumeSecret(VolumeSecretDefault, secretDefault.ObjectMeta.Name),
-						generate.VolumeSecret(VolumeSecretOverride, secretOverride.ObjectMeta.Name),
+						generate.VolumeSecret(VolumeConfigData, configMapData.ObjectMeta.Name),
```
Comment (acbramley) [inline internal/controller/app/drupal/controller.go]: Should be `generate.VolumeConfigMap`

**Pull request merged**